includes/class-wc-checkout.php

```diff
@@ -685,28 +685,37 @@ public function get_posted_data() {
 				continue;
 			}
 
+			$is_checkout_page = isset( $_REQUEST['woocommerce-process-checkout-nonce'] );
+
 			foreach ( $fieldset as $key => $field ) {
 				$type = sanitize_title( isset( $field['type'] ) ? $field['type'] : 'text' );
 
-				// phpcs:disable WordPress.Security.NonceVerification.Missing
-				switch ( $type ) {
-					case 'checkbox':
-						$value = isset( $_POST[ $key ] ) ? 1 : '';
-						break;
-					case 'multiselect':
-						$value = isset( $_POST[ $key ] ) ? implode( ', ', wc_clean( wp_unslash( $_POST[ $key ] ) ) ) : '';
-						break;
-					case 'textarea':
-						$value = isset( $_POST[ $key ] ) ? wc_sanitize_textarea( wp_unslash( $_POST[ $key ] ) ) : '';
-						break;
-					case 'password':
-						$value = isset( $_POST[ $key ] ) ? wp_unslash( $_POST[ $key ] ) : ''; // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
-						break;
-					default:
-						$value = isset( $_POST[ $key ] ) ? wc_clean( wp_unslash( $_POST[ $key ] ) ) : '';
-						break;
+				if ( isset( $_POST[ $key ] ) && '' !== $_POST[ $key ] ) { // phpcs:disable WordPress.Security.NonceVerification.Missing
```

Comment:
I think we should consider the case where we have a **non-required** text field set to an empty string. For instance, consider a custom field defined something like this:

```php
add_filter( 'woocommerce_checkout_fields', function ( $fields ) {
	$fields['billing']['custom_text'] = [
		'type'     => 'text',
		'required' => false,
		'label'    => 'custom text field',
		'default'  => 'this is the default',
	];

	return $fields;
} );
```

If a customer proceeds to checkout they will see the field and it will be pre-populated with the default text. However, it's not _required_ and they may therefore choose to delete it.

* Currently [(trunk@5493218)](https://github.com/woocommerce/woocommerce/commit/5493218870b1234162e20b59f02caf5f4d33c277) if the customer chooses to wipe this field, that's fine.
* With this change in place, however, we effectively set it back to the default any time the submitted value is an empty string.

I think the current logic probably makes sense when we're dealing with required fields, but for non-required fields it will potentially cause breakages with existing third party and custom solutions.

Reply:
> ...we should consider the case where we have a non-required text field set to an empty string...

> ...If a customer proceeds to checkout they will see the field and it will be pre-populated with the default text. However, it's not required and they may therefore choose to delete it.

I think the case you mention it's the same case I mentioned in my comment above but with text inputs https://github.com/woocommerce/woocommerce/pull/29820#discussion_r630648672 

>There was a small issue I've found though. When I added an optional checkbox with "1" as default value. It was saved even though the user unchecked the input. It made sense because the script didn't know if the user unchecked the input or it was never not presented to the user taking its default value instead. 

The suggested solution was to skip checkboxes from the default mechanism. But here I guess we should save the text input as empty string. In that case I'll need the `$is_checkout_page` I created to know if the user was presented with the form or not. Does that sound good to you @barryhughes?

Reply:
@barryhughes I've tested my last commit 79db2d9 and here's the results. I used the following code.

```js
add_filter( 'woocommerce_checkout_fields', function ( $fields ) {
  $fields['billing']['billing_custom_text'] = [
    'type'     => 'text',
    'required' => false,
    'label'    => 'custom text field',
    'default'  => 'this is the default',
  ];

  return $fields;
} );
```

- Case 1. Checkout page, No changes to input.

```
MariaDB [wordpress-one]> SELECT * FROM `wp_postmeta` WHERE `post_id` = 96 AND `meta_key` = '_billing_custom_text' ORDER BY meta_key;
+---------+---------+----------------------+---------------------+
| meta_id | post_id | meta_key             | meta_value          |
+---------+---------+----------------------+---------------------+
|    4880 |      96 | _billing_custom_text | this is the default |
+---------+---------+----------------------+---------------------+
1 row in set (0.000 sec)
```

- Case 2. Checkout page, Input value was changed to "foo"


```
MariaDB [wordpress-one]> SELECT * FROM `wp_postmeta` WHERE `post_id` = 97 AND `meta_key` = '_billing_custom_text' ORDER BY meta_key;
+---------+---------+----------------------+------------+
| meta_id | post_id | meta_key             | meta_value |
+---------+---------+----------------------+------------+
|    4932 |      97 | _billing_custom_text | foo        |
+---------+---------+----------------------+------------+
```

- Case 3. Checkout page, Input value changed to empty string ("")

```
MariaDB [wordpress-one]> SELECT * FROM `wp_postmeta` WHERE `post_id` = 99 AND `meta_key` = '_billing_custom_text' ORDER BY meta_key;
+---------+---------+----------------------+------------+
| meta_id | post_id | meta_key             | meta_value |
+---------+---------+----------------------+------------+
|    5036 |      99 | _billing_custom_text |            |
+---------+---------+----------------------+------------+
1 row in set (0.000 sec)
```


- Case 4. Product page, purchase using Payment Request button


```
MariaDB [wordpress-one]> SELECT * FROM `wp_postmeta` WHERE `post_id` = 98 AND `meta_key` = '_billing_custom_text' ORDER BY meta_key;
+---------+---------+----------------------+---------------------+
| meta_id | post_id | meta_key             | meta_value          |
+---------+---------+----------------------+---------------------+
|    4984 |      98 | _billing_custom_text | this is the default |
+---------+---------+----------------------+---------------------+
1 row in set (0.000 sec)
```

For the case where the input is required, the same behavior is applied except Case 3 is not possible because validation responds with an error message.

Reply:
OK, I think this is good (also seems to work with [Express Checkout](https://woocommerce.com/products/express-checkout-for-woocommerce/) and [One Page Checkout](https://docs.woocommerce.com/document/woocommerce-one-page-checkout/) (I wanted to check those out, as I was curious about compatibility with add-ons). @Konamiman — any other thoughts on this from your end?

Reply:
Nope, I think this behavior is good to go.